File: TypeScript/Algorithm/Algorithm II/11. Container With Most Water.ts
/**
 * Given an array of heights, this function calculates the maximum area of a container
 * that can be formed by selecting two bars and the area between them.
 *
 * @param height - An array of heights representing the bars.
 * @returns The maximum area of a container that can be formed.
 */
function maxArea(height: number[]): number {
  const HEIGHT_LENGTH: number = height.length;

  let maxContainer: number = 0;

  let leftPointer: number = 0;
  let rightPointer: number = HEIGHT_LENGTH - 1;

  while (leftPointer < rightPointer) {
    const leftBar: number = height[leftPointer];
    const rightBar: number = height[rightPointer];
    const barDistance: number = rightPointer - leftPointer;

    const currentContainer: number = Math.min(leftBar, rightBar) * barDistance;
    maxContainer = Math.max(maxContainer, currentContainer);

    if (leftBar < rightBar) leftPointer++;
    else rightPointer--;
  }

  return maxContainer;
}

File: Java/Algorithm/Algorithm II/11. Container With Most Water.java
class Solution {

  public int maxArea(int[] height) {
    int HEIGHT_LENGTH = height.length;

    int maxContainer = 0;

    int leftPointer = 0;
    int rightPointer = HEIGHT_LENGTH - 1;

    while (leftPointer < rightPointer) {
      int leftBar = height[leftPointer];
      int rightBar = height[rightPointer];
      int barDistance = rightPointer - leftPointer;

      int currentContainer = Math.min(leftBar, rightBar) * barDistance;
      maxContainer = Math.max(maxContainer, currentContainer);

      if (leftBar < rightBar) leftPointer++;
      else rightPointer--;
    }

    return maxContainer;
  }
}
